### ISSUE TITLE
Mismatching between custom LabelSmoothing and and PyTorch's Label Smoothing

### ISSUE THREAD

**Issue by csukuangfj** [k2-fsa/icefall#107]
The current LabelSmooting in icefall (see below)
https://github.com/k2-fsa/icefall/blob/810b193dcc3ad3f7a65bc3def63493711c9a084e/egs/librispeech/ASR/conformer_ctc/transformer.py#L797

is based on ESPnet, which seems to be based on [The Annotated Transformer](https://nlp.seas.harvard.edu/2018/04/03/attention.html#label-smoothing).

As @janvainer  pointed out in #106, there is a built-in LabelSmooting loss in torch >=1.10, see https://pytorch.org/docs/master/generated/torch.nn.CrossEntropyLoss.html#torch.nn.CrossEntropyLoss

I just compared the implementation from PyTorch with the one in icefall and identified the following differences.

PyTorch's implementation follows the one described in the paper [Attention Is All You Need](https://arxiv.org/pdf/1706.03762.pdf)

LabelSmoothing is proposed by the paper [Rethinking the Inception Architecture for Computer Vision](https://arxiv.org/pdf/1512.00567.pdf), which has the following formula:
<img width="201" alt="Screen Shot 2021-11-05 at 7 27 49 PM" src="https://user-images.githubusercontent.com/5284924/140503799-7ec5f93b-1a35-4486-b346-57ee34455f84.png">

### First difference

icefall is using `K - 1`, not `K`. See the code below from icefall
https://github.com/k2-fsa/icefall/blob/810b193dcc3ad3f7a65bc3def63493711c9a084e/egs/librispeech/ASR/conformer_ctc/transformer.py#L853

### Second difference

 [Rethinking the Inception Architecture for Computer Vision](https://arxiv.org/pdf/1512.00567.pdf) uses cross-entropy to compute the loss. See the formula from the paper, given below:
<img width="243" alt="Screen Shot 2021-11-05 at 7 32 21 PM" src="https://user-images.githubusercontent.com/5284924/140504304-17945815-b685-449c-848c-e63d0cf80026.png">

but icefall uses the following formula, i.e., [KL-divergence](https://pytorch.org/docs/master/generated/torch.nn.KLDivLoss.html#torch.nn.KLDivLoss)
![CodeCogsEqn](https://user-images.githubusercontent.com/5284924/140504914-7ba6d7f8-3005-47f0-9404-8f938edabcf5.gif)

---

To match PyTorch's implementation (also the one used in the original transformer paper), we have to do the following changes:

(1) Change
https://github.com/k2-fsa/icefall/blob/810b193dcc3ad3f7a65bc3def63493711c9a084e/egs/librispeech/ASR/conformer_ctc/transformer.py#L853
to
```python3
true_dist.fill_(self.smoothing / self.size)
```

(2) Change
https://github.com/k2-fsa/icefall/blob/810b193dcc3ad3f7a65bc3def63493711c9a084e/egs/librispeech/ASR/conformer_ctc/transformer.py#L858
to

```python3
lable_smoothing_loss = -1 * torch.log_softmax(x, dim=1) * true_dist.sum(dim=1)
```

---
@danpovey  Do you think we should make the above changes?

**Comment by danpovey** [comment]
Yes I agree that it would be simpler and easier to understand.

**Comment by janvainer** [comment]
Thanks for doing the comparison!
